Protect Segment Store metadata against Zookeeper client inconsistent replies

**Problem description**
There is general issue in the Segment Store when dealing with Zookeeper metadata under network/connectivity problems. In a nutshell, the `BookkeeperLog` class, which is intended to manage the metadata Bookkeeper ledgers in Zookeeper, can get a `BadVersionException`, which is expected to fail the write to Zookeeper, but the data has been actually written. This leads the Zookeeper metadata managed by the Segment Store to be in an inconsistent state. This is a summary of one instance of this problem:

1) The problems start around "2020-04-16 02:59:32", in which Segment Store 0 has problems connecting to Zookeeper:
```
2020-04-16 02:59:32,369 1812318246 [ZKC-connect-executor-0-EventThread] INFO o.a.b.z.ZooKeeperWatcherBase - ZooKeeper client is disconnected from zookeeper now, but it is OK unless we received EXPIRED event.
...
2020-04-16 02:59:33,771 1812319648 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] INFO org.apache.zookeeper.ClientCnxn - Opening socket connection to server nautilus-pravega-zookeeper-client/yyy:2181. Will not attempt to authenticate using SASL (unknown error)
2020-04-16 02:59:33,772 1812319649 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] INFO org.apache.zookeeper.ClientCnxn - Socket connection established, initiating session, client: /xxx:60894, server: nautilus-pravega-zookeeper-client/zzz:2181
2020-04-16 02:59:33,774 1812319651 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] WARN org.apache.zookeeper.ClientCnxn - Unable to reconnect to ZooKeeper service, session 0x3005595379a7d2e has expired
2020-04-16 02:59:33,774 1812319651 [ZKC-connect-executor-0-EventThread] INFO o.a.b.meta.AbstractZkLedgerManager - ZooKeeper client expired on ledger manager.
2020-04-16 02:59:33,774 1812319651 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] INFO org.apache.zookeeper.ClientCnxn - Unable to reconnect to ZooKeeper service, session 0x3005595379a7d2e has expired, closing socket connection
2020-04-16 02:59:33,774 1812319651 [ZKC-connect-executor-0-EventThread] ERROR o.a.b.z.ZooKeeperWatcherBase - ZooKeeper client connection to the ZooKeeper server has expired!
2020-04-16 02:59:33,775 1812319652 [ZKC-connect-executor-0-EventThread] INFO o.a.b.zookeeper.ZooKeeperClient - ZooKeeper session 3005595379a7d2e is expired from nautilus-pravega-zookeeper-client:2181.
2020-04-16 02:59:33,775 1812319652 [ZKC-connect-executor-0-EventThread] INFO org.apache.zookeeper.ClientCnxn - EventThread shut down for session: 0x3005595379a7d2e
2020-04-16 02:59:33,777 1812319654 [ZKC-connect-executor-0] INFO o.a.b.zookeeper.ZooKeeperClient - Reconnecting zookeeper nautilus-pravega-zookeeper-client:2181.
```

2) Around this time, we observe that a new ledger is created (`1090530`) for Container 0 in Segment Store 0. Note that this snippet is from 30 seconds after the previous one, and there still are problems connecting to Zookeeper:
```
2020-04-16 02:59:46,985 1812332862 [ZKC-connect-executor-0-EventThread] INFO o.a.b.client.LedgerCreateOp - Ensemble: [nautilus-bookie-2.nautilus-bookie-headless.nautilus-pravega.svc.cluster.local:3181, nautilus-bookie-4.nautilus-bookie-headless.nautilus-pravega.svc.cluster.local:3181, nautilus-bookie-1.nautilus-bookie-headless.nautilus-pravega.svc.cluster.local:3181] for ledger: 1090530
2020-04-16 02:59:46,985 1812332862 [core-19] INFO i.p.s.s.i.bookkeeper.BookKeeperLog - Log[0]: Created Ledger 1090530.
2020-04-16 02:59:59,472 1812345349 [core-2] INFO i.p.s.server.CacheManager - CacheManager Registered io.pravega.segmentstore.server.tables.ContainerKeyCache@27f53aab.
2020-04-16 02:59:59,472 1812345349 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] WARN org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 12487ms for sessionid 0x2006ed9c323157e
2020-04-16 02:59:59,472 1812345349 [ZKC-connect-executor-0-SendThread(nautilus-pravega-zookeeper-client:2181)] INFO org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 12487ms for sessionid 0x2006ed9c323157e, closing socket connection and attempting reconnect
```

3) After creating the Ledger, the Segment Store tries to update the metadata of ledgers in Zookeeper, but it **fails (?)** because "apparently" it has been fenced out by another Segment Store:
```
2020-04-16 03:00:04,610 1812350487 [core-19] ERROR i.p.s.server.logs.DurableLog - DurableLog[0] Recovery FAILED.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/nautilus/segmentstore/containers/0/0/0').
...
Caused by: org.apache.zookeeper.KeeperException$BadVersionException: KeeperErrorCode = BadVersion for /pravega/nautilus/segmentstore/containers/0/0/0
```
In any case, getting a `BadVersionException` means that the **addition of ledger `1090530` to the Zookeeper metadata has failed (i.e., not committed)**.

4) This is strange, because in theory Segment Store 0 owns Container 0, but we can see that there is some churn in the ownership of containers during this time:
```
Segment Store 0
2020-04-16 02:56:27,286 1812133163 [core-12] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [16, 1, 33, 17, 34, 2, 19, 22], Current = [16, 1, 17, 33, 2, 34, 19, 22], PendingTasks = [], ToStart = [], ToStop = []. 2020-04-16 02:57:41,368 1812207245 [core-12] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [16, 32, 1, 33, 17, 34, 2, 19, 21, 22], Current = [16, 1, 17, 33, 2, 34, 19, 22], PendingTasks = [], ToStart = [32, 21], ToStop = [].
2020-04-16 02:58:11,741 1812237618 [core-26] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [1, 33, 17, 34, 2, 19, 21, 22], Current = [16, 32, 1, 17, 33, 2, 34, 19, 21, 22], PendingTasks = [], ToStart = [], ToStop = [16, 32].
2020-04-16 02:59:34,186 1812320063 [core-28] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [], Current = [1, 17, 33, 2, 34, 19, 21, 22], PendingTasks = [], ToStart = [], ToStop = [1, 17, 33, 2, 34, 19, 21, 22]. 2020-04-16 02:59:46,971 1812332848 [core-2] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 17, 33, 2, 34, 19, 21], Current = [], PendingTasks = [], ToStart = [0, 16, 17, 33, 2, 34, 19, 21], ToStop = [].
2020-04-16 02:59:59,498 1812345375 [core-7] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 17, 33, 2, 34, 19, 21], Current = [], PendingTasks = [0, 16, 17, 33, 2, 34, 19, 21], ToStart = [], ToStop = [].
```
```
Segment Store 3
2020-04-16 02:46:35,841 1811466360 [core-20] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [32, 0, 18, 3, 35, 21, 37, 27], Current = [32, 0, 18, 35, 3, 37, 21, 27], PendingTasks = [], ToStart = [], ToStop = []. 
2020-04-16 02:56:43,468 1812073987 [core-9] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [32, 0, 18, 3, 35, 21, 37, 27], Current = [32, 0, 18, 35, 3, 37, 21, 27], PendingTasks = [], ToStart = [], ToStop = [].
2020-04-16 02:58:02,150 1812152669 [core-3] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [], Current = [32, 0, 18, 35, 3, 37, 21, 27], PendingTasks = [], ToStart = [], ToStop = [32, 0, 18, 35, 3, 37, 21, 27].
2020-04-16 02:58:14,091 1812164610 [core-28] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 18, 35, 3, 4, 27], Current = [27], PendingTasks = [27], ToStart = [0, 16, 32, 18, 35, 3, 4], ToStop = [].
2020-04-16 02:58:26,058 1812176577 [core-26] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 18, 35, 3, 4, 27], Current = [], PendingTasks = [0, 16, 32, 18, 3, 35, 4], ToStart = [27], ToStop = [].
2020-04-16 02:58:37,624 1812188143 [core-20] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 18, 35, 3, 4, 27], Current = [], PendingTasks = [0, 16, 32, 18, 3, 35, 4, 27], ToStart = [], ToStop = [].
2020-04-16 02:58:47,625 1812198144 [core-2] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 18, 35, 3, 4, 27], Current = [], PendingTasks = [0, 16, 32, 18, 3, 35, 4, 27], ToStart = [], ToStop = [].
2020-04-16 02:58:57,625 1812208144 [core-11] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 18, 35, 3, 4, 27], Current = [32, 16, 35, 3, 4, 27], PendingTasks = [0, 18], ToStart = [], ToStop = [].
2020-04-16 02:59:22,388 1812232907 [core-14] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 16, 32, 1, 18, 35, 3, 4, 22, 27], Current = [32, 0, 16, 18, 35, 3, 4, 27], PendingTasks = [], ToStart = [1, 22], ToStop = [].
2020-04-16 02:59:46,968 1812257487 [core-13] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [32, 1, 18, 35, 3, 4, 22, 27], Current = [32, 0, 16, 1, 18, 35, 3, 4, 22, 27], PendingTasks = [], ToStart = [], ToStop = [0, 16].
2020-04-16 03:09:47,934 1812858453 [core-21] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [32, 1, 18, 35, 3, 4, 22, 27], Current = [32, 1, 18, 35, 3, 4, 22, 27], PendingTasks = [], ToStart = [], ToStop = [].
```
```
Segment Store 4
2020-04-16 02:56:28,561 1812031775 [core-10] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [4, 20, 36, 7, 26, 10, 43, 30], Current = [4, 36, 20, 7, 26, 10, 43, 30], PendingTasks = [], ToStart = [], ToStop = []. 
2020-04-16 02:57:40,461 1812103675 [core-9] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 4, 20, 36, 37, 7, 26, 10, 43, 30], Current = [4, 36, 20, 7, 26, 10, 43, 30], PendingTasks = [], ToStart = [0, 37], ToStop = [].
2020-04-16 02:58:11,704 1812134918 [core-13] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [20, 36, 37, 7, 26, 10, 43, 30], Current = [0, 4, 36, 20, 37, 7, 26, 10, 43, 30], PendingTasks = [], ToStart = [], ToStop = [0, 4].
2020-04-16 02:59:22,363 1812205577 [core-10] INFO i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [33, 20, 36, 37, 7, 26, 10, 43, 30], Current = [36, 20, 37, 7, 26, 10, 43, 30], PendingTasks = [], ToStart = [33], ToStop = [].
``` 

5) In any case, once the Segment Store gets a `BadVersionException` when attempting to update the Zookeeper metadata in the `BookkeeperLog:initialize()` method, the recently created ledger `1090530` is correctly deleted from Bookkeeper. This works fine.

6) But here is the problem: the ledger `1090530` created for Container 0 has been correctly deleted from Bookkeeper, **but it appears in the Zookeeper metadata managed by the Segment Store**. This means that the `BadVersionException` got from the Zookeeper client (Curator), actually wrote the ledger id in Zookeeper. The consequence of this is that when the Container tries to recover, it complains given that there is a ledger in the Zookeeper metadata that does not exists in Bookkeeper:

```
2020-04-16 03:00:08,362 1812354239 [core-28] INFO i.p.s.s.c.StreamSegmentContainer - SegmentContainer[0]: Starting.
2020-04-16 03:00:08,362 1812354239 [core-28] INFO i.p.s.server.logs.DurableLog - DurableLog[0]: Starting.
2020-04-16 03:00:08,368 1812354245 [core-29] ERROR i.p.s.server.logs.DurableLog - DurableLog[0] Recovery FAILED.
io.pravega.segmentstore.storage.DurableDataLogException: Unable to open-fence ledger 1090530.
...
Caused by: org.apache.bookkeeper.client.BKException$BKNoSuchLedgerExistsException: No such ledger exists
...
2020-04-16 03:03:09,555 1812535432 [core-5] INFO i.p.s.s.c.StreamSegmentContainer - SegmentContainer[0]: Starting.
2020-04-16 03:03:09,555 1812535432 [core-5] INFO i.p.s.server.logs.DurableLog - DurableLog[0]: Starting.
2020-04-16 03:03:09,562 1812535439 [core-4] ERROR i.p.s.server.logs.DurableLog - DurableLog[0] Recovery FAILED.
io.pravega.segmentstore.storage.DurableDataLogException: Unable to open-fence ledger 1090530.
...
Caused by: org.apache.bookkeeper.client.BKException$BKNoSuchLedgerExistsException: No such ledger exists
```

This means that, under some concrete circumstances in which network has glitches and the Zookeeper client is attempting to reconnect, we could have a `BadVersionException` response from the client that actually writes the data to Zookeeper. We need to protect the Segment Store against this kind of situations.

**Problem location**
`BookkeeperLog`, and any other place in which metadata in Zookeeper is updated by the Segment Store.

**Suggestions for an improvement**
If it is not possible to find and get a fix for the Curator client, we need to add some defensive code to prevent this situation.

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java
@@ -812,7 +812,11 @@ class BookKeeperLog implements DurableDataLog {
 
         try {
             persistMetadata(currentMetadata, create);
-        } catch (DurableDataLogException ex) {
+        } catch (DataLogWriterNotPrimaryException ex) {
+            // Only attempt to cleanup the newly created ledger if we were fenced out. Any other exception is not indicative
+            // of whether we were able to persist the metadata or not, so it's safer to leave the ledger behind in case
+            // it is still used. If indeed our metadata has been updated, a subsequent recovery will pick it up and delete it
+            // because it (should be) empty.
             try {
                 Ledgers.delete(newLedger.getId(), this.bookKeeper);
             } catch (Exception deleteEx) {
@@ -820,6 +824,9 @@ class BookKeeperLog implements DurableDataLog {
                 ex.addSuppressed(deleteEx);
             }
 
+            throw ex;
+        } catch (Exception ex) {
+            log.warn("{}: Error while using ZooKeeper. Leaving orphaned ledger {} behind.", this.traceObjectId, newLedger.getId());
             throw ex;
         }
 

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java
@@ -840,34 +840,60 @@ class BookKeeperLog implements DurableDataLog {
     private void persistMetadata(LogMetadata metadata, boolean create) throws DurableDataLogException {
         try {
             byte[] serializedMetadata = LogMetadata.SERIALIZER.serialize(metadata).getCopy();
-            Stat result;
-            if (create) {
-                result = new Stat();
-                this.zkClient.create()
-                             .creatingParentsIfNeeded()
-                             .storingStatIn(result)
-                             .forPath(this.logNodePath, serializedMetadata);
-            } else {
-                result = this.zkClient.setData()
-                                      .withVersion(metadata.getUpdateVersion())
-                                      .forPath(this.logNodePath, serializedMetadata);
-
-            }
-
+            Stat result = create
+                    ? createZkMetadata(serializedMetadata)
+                    : updateZkMetadata(serializedMetadata, metadata.getUpdateVersion());
             metadata.withUpdateVersion(result.getVersion());
         } catch (KeeperException.NodeExistsException | KeeperException.BadVersionException keeperEx) {
-            // We were fenced out. Clean up and throw appropriate exception.
-            throw new DataLogWriterNotPrimaryException(
-                    String.format("Unable to acquire exclusive write lock for log (path = '%s%s').", this.zkClient.getNamespace(), this.logNodePath),
-                    keeperEx);
+            if (reconcileMetadata(metadata)) {
+                log.info("{}: Received '{}' from ZooKeeper while persisting metadata (path = '{}{}'), however metadata has been persisted correctly. Not rethrowing.",
+                        this.traceObjectId, keeperEx.toString(), this.zkClient.getNamespace(), this.logNodePath);
+            } else {
+                // We were fenced out. Convert to an appropriate exception.
+                throw new DataLogWriterNotPrimaryException(
+                        String.format("Unable to acquire exclusive write lock for log (path = '%s%s').", this.zkClient.getNamespace(), this.logNodePath),
+                        keeperEx);
+            }
         } catch (Exception generalEx) {
-            // General exception. Clean up and rethrow appropriate exception.
+            // General exception. Convert to an appropriate exception.
             throw new DataLogInitializationException(
                     String.format("Unable to update ZNode for path '%s%s'.", this.zkClient.getNamespace(), this.logNodePath),
                     generalEx);
         }
 
         log.info("{} Metadata persisted ({}).", this.traceObjectId, metadata);
+    }
+
+    @VisibleForTesting
+    protected Stat createZkMetadata(byte[] serializedMetadata) throws Exception {
+        val result = new Stat();
+        this.zkClient.create().creatingParentsIfNeeded().storingStatIn(result).forPath(this.logNodePath, serializedMetadata);
+        return result;
+    }
+
+    @VisibleForTesting
+    protected Stat updateZkMetadata(byte[] serializedMetadata, int version) throws Exception {
+        return this.zkClient.setData().withVersion(version).forPath(this.logNodePath, serializedMetadata);
+    }
+
+    /**
+     * Verifies the given {@link LogMetadata} against the actual one stored in ZooKeeper.
+     *
+     * @param metadata The Metadata to check.
+     * @return True if the metadata stored in ZooKeeper is an identical match to the given one, false otherwise. If true,
+     * {@link LogMetadata#getUpdateVersion()} will also be updated with the one stored in ZooKeeper.
+     */
+    private boolean reconcileMetadata(LogMetadata metadata) {
+        try {
+            val actualMetadata = loadMetadata();
+            if (metadata.equals(actualMetadata)) {
+                metadata.withUpdateVersion(actualMetadata.getUpdateVersion());
+                return true;
+            }
+        } catch (DataLogInitializationException ex) {
+            log.warn("{}: Unable to verify persisted metadata (path = '{}{}').", this.traceObjectId, this.zkClient.getNamespace(), this.logNodePath, ex);
+        }
+        return false;
     }
 
     /**
@@ -890,6 +916,7 @@ class BookKeeperLog implements DurableDataLog {
 
         persistMetadata(metadata, create);
     }
+
 
     //endregion
 

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LedgerMetadata.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LedgerMetadata.java
@@ -10,6 +10,7 @@
 package io.pravega.segmentstore.storage.impl.bookkeeper;
 
 import lombok.AccessLevel;
+import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 
@@ -18,6 +19,7 @@ import lombok.RequiredArgsConstructor;
  */
 @RequiredArgsConstructor
 @Getter
+@EqualsAndHashCode
 public class LedgerMetadata {
     /**
      * The BookKeeper-assigned Ledger Id.

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ReadOnlyLogMetadata.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ReadOnlyLogMetadata.java
@@ -52,4 +52,35 @@ public interface ReadOnlyLogMetadata {
      * @return The Truncation Address.
      */
     LedgerAddress getTruncationAddress();
+
+    /**
+     * Determines whether this {@link ReadOnlyLogMetadata} is equivalent to the other one.
+     *
+     * @param other The other instance.
+     * @return True if equivalent, false otherwise.
+     */
+    default boolean equals(ReadOnlyLogMetadata other) {
+        if (other == null) {
+            return false;
+        }
+
+        List<LedgerMetadata> ledgers = getLedgers();
+        List<LedgerMetadata> otherLedgers = other.getLedgers();
+        if (this.isEnabled() != other.isEnabled()
+                || this.getEpoch() != other.getEpoch()
+                || !this.getTruncationAddress().equals(other.getTruncationAddress())
+                || ledgers.size() != otherLedgers.size()) {
+            return false;
+        }
+
+        // Check each ledger.
+        for (int i = 0; i < ledgers.size(); i++) {
+            if (!ledgers.get(i).equals(otherLedgers.get(i))) {
+                return false;
+            }
+        }
+
+        // All tests have passed.
+        return true;
+    }
 }

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
@@ -388,6 +388,10 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
 
     @Test
     public void testReconcileMetadata() throws Exception {
+        @Cleanup
+        BookKeeperAdmin a = new BookKeeperAdmin(this.factory.get().getBookKeeperClient());
+        val initialLedgers = Sets.newHashSet(a.listLedgers());
+
         // Test initialization (node creation).
         try (val log = new TestBookKeeperLog()) {
             // Data not persisted and we throw an error - this is a real fencing event.
@@ -425,9 +429,8 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
         }
 
         // Verify ledger cleanup.
-        @Cleanup
-        BookKeeperAdmin a = new BookKeeperAdmin(this.factory.get().getBookKeeperClient());
         val allLedgers = Sets.newHashSet(a.listLedgers());
+        allLedgers.removeAll(initialLedgers);
         Assert.assertEquals("Unexpected ledgers in BK.", expectedLedgerIds, allLedgers);
     }
 

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
@@ -10,6 +10,7 @@
 package io.pravega.segmentstore.storage.impl.bookkeeper;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
 import io.pravega.common.ObjectClosedException;
 import io.pravega.common.concurrent.Futures;
 import io.pravega.common.util.CompositeByteArraySegment;
@@ -31,6 +32,7 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
@@ -48,6 +50,7 @@ import lombok.SneakyThrows;
 import lombok.val;
 import org.apache.bookkeeper.client.BKException;
 import org.apache.bookkeeper.client.BookKeeper;
+import org.apache.bookkeeper.client.BookKeeperAdmin;
 import org.apache.bookkeeper.client.LedgerHandle;
 import org.apache.curator.framework.CuratorFramework;
 import org.apache.curator.framework.CuratorFrameworkFactory;
@@ -402,6 +405,7 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
             Assert.assertEquals("Create(Persist=True, Throw=True)", 2, log.getCreateExceptionCount());
         }
 
+        val expectedLedgerIds = new HashSet<Long>();
         // Test updates (subsequent recoveries).
         try (val log = new TestBookKeeperLog()) {
             // Data not persisted and we throw an error - this is a real fencing event.
@@ -417,7 +421,14 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
             log.setPersistData(true);
             log.initialize(TIMEOUT);
             Assert.assertEquals("Update(Persist=True, Throw=True)", 2, log.getUpdateExceptionCount());
+            log.loadMetadata().getLedgers().stream().map(LedgerMetadata::getLedgerId).forEach(expectedLedgerIds::add);
         }
+
+        // Verify ledger cleanup.
+        @Cleanup
+        BookKeeperAdmin a = new BookKeeperAdmin(this.factory.get().getBookKeeperClient());
+        val allLedgers = Sets.newHashSet(a.listLedgers());
+        Assert.assertEquals("Unexpected ledgers in BK.", expectedLedgerIds, allLedgers);
     }
 
     /**

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogTests.java
@@ -9,12 +9,14 @@
  */
 package io.pravega.segmentstore.storage.impl.bookkeeper;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.pravega.common.ObjectClosedException;
 import io.pravega.common.concurrent.Futures;
 import io.pravega.common.util.CompositeByteArraySegment;
 import io.pravega.common.util.RetriesExhaustedException;
 import io.pravega.segmentstore.storage.DataLogCorruptedException;
 import io.pravega.segmentstore.storage.DataLogNotAvailableException;
+import io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException;
 import io.pravega.segmentstore.storage.DurableDataLog;
 import io.pravega.segmentstore.storage.DurableDataLogException;
 import io.pravega.segmentstore.storage.DurableDataLogTestBase;
@@ -41,6 +43,7 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import lombok.Cleanup;
+import lombok.Setter;
 import lombok.SneakyThrows;
 import lombok.val;
 import org.apache.bookkeeper.client.BKException;
@@ -49,6 +52,8 @@ import org.apache.bookkeeper.client.LedgerHandle;
 import org.apache.curator.framework.CuratorFramework;
 import org.apache.curator.framework.CuratorFrameworkFactory;
 import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
 import org.junit.After;
 import org.junit.AfterClass;
 import org.junit.Assert;
@@ -376,6 +381,43 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
         val ws = log.getWriteSettings();
         Assert.assertEquals(BookKeeperConfig.MAX_APPEND_LENGTH, ws.getMaxWriteLength());
         Assert.assertEquals(this.config.get().getMaxOutstandingBytes(), ws.getMaxOutstandingBytes());
+    }
+
+    @Test
+    public void testReconcileMetadata() throws Exception {
+        // Test initialization (node creation).
+        try (val log = new TestBookKeeperLog()) {
+            // Data not persisted and we throw an error - this is a real fencing event.
+            log.setThrowZkException(true);
+            log.setPersistData(false);
+            AssertExtensions.assertThrows(
+                    "Create(Persist=False, Throw=True)",
+                    () -> log.initialize(TIMEOUT),
+                    ex -> ex instanceof DataLogWriterNotPrimaryException);
+            Assert.assertEquals(1, log.getCreateExceptionCount());
+
+            // Data persisted correctly and we throw an error - reconciliation needed.
+            log.setPersistData(true);
+            log.initialize(TIMEOUT);
+            Assert.assertEquals("Create(Persist=True, Throw=True)", 2, log.getCreateExceptionCount());
+        }
+
+        // Test updates (subsequent recoveries).
+        try (val log = new TestBookKeeperLog()) {
+            // Data not persisted and we throw an error - this is a real fencing event.
+            log.setThrowZkException(true);
+            log.setPersistData(false);
+            AssertExtensions.assertThrows(
+                    "Update(Persist=False, Throw=True)",
+                    () -> log.initialize(TIMEOUT),
+                    ex -> ex instanceof DataLogWriterNotPrimaryException);
+            Assert.assertEquals(1, log.getUpdateExceptionCount());
+
+            // Data persisted correctly and we throw an error - reconciliation needed.
+            log.setPersistData(true);
+            log.initialize(TIMEOUT);
+            Assert.assertEquals("Update(Persist=True, Throw=True)", 2, log.getUpdateExceptionCount());
+        }
     }
 
     /**
@@ -738,6 +780,61 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
     }
 
     //endregion
+
+    //region TestBookKeeperLog
+
+    @Setter
+    private class TestBookKeeperLog extends BookKeeperLog {
+        private boolean persistData = true;
+        private boolean throwZkException = false;
+        private final AtomicInteger createExceptionCount = new AtomicInteger();
+        private final AtomicInteger updateExceptionCount = new AtomicInteger();
+
+        TestBookKeeperLog() {
+            super(CONTAINER_ID, zkClient.get(), factory.get().getBookKeeperClient(), config.get(), executorService());
+        }
+
+        int getCreateExceptionCount() {
+            return this.createExceptionCount.get();
+        }
+
+        int getUpdateExceptionCount() {
+            return this.updateExceptionCount.get();
+        }
+
+        @Override
+        protected Stat createZkMetadata(byte[] serializedMetadata) throws Exception {
+            assert this.persistData || this.throwZkException;
+            Stat result = new Stat();
+            if (this.persistData) {
+                result = super.createZkMetadata(serializedMetadata);
+            }
+            if (this.throwZkException) {
+                this.createExceptionCount.incrementAndGet();
+                throw new KeeperException.NodeExistsException();
+            }
+            return result;
+        }
+
+        @VisibleForTesting
+        protected Stat updateZkMetadata(byte[] serializedMetadata, int version) throws Exception {
+            assert this.persistData || this.throwZkException;
+            Stat result = new Stat();
+            if (this.persistData) {
+                super.updateZkMetadata(serializedMetadata, version);
+            }
+            if (this.throwZkException) {
+                this.updateExceptionCount.incrementAndGet();
+                throw new KeeperException.BadVersionException();
+            }
+            return result;
+        }
+    }
+
+    //endregion
+
+    //region Actual Test Implementations
+
     public static class SecureBookKeeperLogTests extends BookKeeperLogTests {
         @BeforeClass
         public static void startUp() throws Exception {
@@ -751,4 +848,6 @@ public abstract class BookKeeperLogTests extends DurableDataLogTestBase {
             setUpBookKeeper(false);
         }
     }
+
+    //endregion
 }

--- a/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogMetadataTests.java
+++ b/segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogMetadataTests.java
@@ -11,6 +11,8 @@ package io.pravega.segmentstore.storage.impl.bookkeeper;
 
 import io.pravega.test.common.AssertExtensions;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicLong;
@@ -216,6 +218,51 @@ public class LogMetadataTests {
         Assert.assertEquals("Unexpected TruncationAddress.", m1.getTruncationAddress().getLedgerId(), m2.getTruncationAddress().getLedgerId());
         AssertExtensions.assertListEquals("Unexpected ledgers.", m1.getLedgers(), m2.getLedgers(),
                 (l1, l2) -> l1.getSequence() == l2.getSequence() && l1.getLedgerId() == l2.getLedgerId() && l1.getStatus() == l2.getStatus());
+    }
+
+    /**
+     * Tests {@link ReadOnlyLogMetadata#equals}.
+     */
+    @Test
+    public void testEquals() {
+        val m1 = LogMetadata.builder()
+                .enabled(true)
+                .epoch(1)
+                .updateVersion(2)
+                .truncationAddress(new LedgerAddress(100L, 100L))
+                .ledgers(Arrays.asList(new LedgerMetadata(1000L, 100, LedgerMetadata.Status.Unknown),
+                        new LedgerMetadata(2000L, 200, LedgerMetadata.Status.NotEmpty)))
+                .build();
+
+        // Check against null.
+        Assert.assertFalse(m1.equals(null));
+
+        // A copy of itself should be identical.
+        Assert.assertTrue(m1.equals(copyOf(m1)));
+
+        // Enabled/disabled.
+        Assert.assertFalse(m1.equals(m1.asDisabled()));
+        Assert.assertTrue(m1.equals(m1.asDisabled().asEnabled()));
+
+        // Update version is not part of the equality check.
+        Assert.assertTrue(m1.equals(m1.withUpdateVersion(234)));
+
+        // Truncation address.
+        Assert.assertFalse(m1.equals(m1.truncate(new LedgerAddress(200L, 100L))));
+
+        // Ledgers.
+        Assert.assertFalse(m1.equals(m1.updateLedgerStatus(Collections.singletonMap(1000L, Ledgers.NO_ENTRY_ID + 1))));
+        Assert.assertFalse(m1.equals(m1.addLedger(300)));
+    }
+
+    private LogMetadata copyOf(LogMetadata m) {
+        return LogMetadata.builder()
+                .enabled(m.isEnabled())
+                .epoch(m.getEpoch())
+                .updateVersion(m.getUpdateVersion())
+                .truncationAddress(new LedgerAddress(m.getTruncationAddress().getSequence(), m.getTruncationAddress().getLedgerSequence()))
+                .ledgers(m.getLedgers().stream().map(lm -> new LedgerMetadata(lm.getLedgerId(), lm.getSequence(), lm.getStatus())).collect(Collectors.toList()))
+                .build();
     }
 
     private void checkLedgerIds(List<Long> expectedLedgerIds, LogMetadata metadata) {

